edgedb-tokio/src/query_executor.rs

```diff
@@ -0,0 +1,192 @@
+use edgedb_protocol::model::Json;
+use edgedb_protocol::query_arg::QueryArgs;
+use edgedb_protocol::QueryResult;
+use std::future::Future;
+
+use crate::{Client, Error, Transaction};
+
+/// Abstracts over different query executors
+/// In particular &Client and &mut Transaction
+pub trait QueryExecutor {
+    /// see [Client::query]
+    fn query<R, A>(self, query: &str, arguments: &A) -> impl Future<Output = Result<Vec<R>, Error>>
```

Comment:
> Handling of `Send` needs to be discussed. Currently I'm not constraining anything to `Send`, but perhaps the returned futures need to be, but that might require adding constraints to some of the generic arguments (e.g. `QueryResult`) as well.

Reply:
I think this future should be required to be `Send`.

Imagine that for some reason, the query results contains a `Rc`. Imagine that during poll of the future this `Rc` is somehow mutated & moved across threads. That is not ok.

This is obscure and unlikely to happen and also not a concern of this PR - it should be `QueryResult` that enforces this constraint.

Reply:
I think adding `+Sync` to the return value requires adding `+Sync` on `QueryResult`, or it won't compile. I can easily make that change. 

But for some reason neither `Client` nor `Transaction` had that constraint on `QueryResult` (which isn't a problem per-se, since the trait having stricter constraints isn't a problem). Does that mean that `Client`/`Transaction` only work on single threaded executors?

But I have little experience with async, so I don't fully understand how `Sync` interacts with it.

Reply:
> I think adding `+Sync`

You mean `Send`, right?

> Does that mean that Client/Transaction only work on single threaded executors?

No, it means that the future that these function return can be sent between threads. It is a requirement that these function are implemented in a way that can be multi-threaded.

This requirement is a good thing, but it could be too strict for some cases.

Reply:
> You mean Send, right?

yes

I am confused how the methods on `Client`/`Transaction` work without constraining their inputs/output types to `Send`. But I guess it can work because they finish serializing the inputs before going async, and only start deserializing the outputs after finished the async part? But doesn't input serialization require fetching type descriptors asynchronously? So how can the input be kept across that async, while keeping the returned future `Send`? Or is it simply inferred so that the future is `Send` if the arguments are `Send`, and arguments which aren't `Send` mean the returned future isn't `Send`?

Reply:
That's right: for concrete types, Rust compiler can infer if they are `Send` (and `Sync`) or not. It's only for generics or dyn types that you need to add an explicit `Send`.